### ISSUE TITLE
Info: Crash when loading any linux vst in Carla

### ISSUE THREAD

**Issue by RobinHerbots** [falkTX/Carla#1000]
Hi, 

I installed Carla on Ubuntu 19.10 with the kx PPA. 
But Carla seems to crash whenever I do addplugin and select a native linux VST.  It just closes carla.

Probably it is a configuration problem on my side.  Is there some logs or crashdump so I can get a clue why this is happening.  Console out just tells that Carla has closed.

Beste regards, 
Robin

**Comment by wundermusiker** [comment]
Same with me. KX PPA Ubuntu 18.04.
carla[26628]: segfault at 0 ip 00007feb454ad128 sp 00007feb300f9d10 error 4 in libcarla_standalone2.so[7feb453d6000+b6d000]

**Comment by falkTX** [comment]
I cant reproduce here. Can you try doing a full plugin rescan?

**Comment by RobinHerbots** [comment]
@falkTX , 

Althought, I added the vst in to the paths to scan, carla doesn't seem to detect them.  When I add them by selecting the vst file Carla closes and I get 
Illegal instruction (core dumped)

Will it be the plugins itself?  I tried Dexed and Tunefish4.
Does carla load windows vst?

**Comment by wundermusiker** [comment]
If you can tell me how. Sure. I tried to reproduce. It happens only when I start in this order:
1. Cadence-Ardour-Hydrogen-Carla-Carla loading any plugin crashes.
1. Cadence-Carla-Carla loading Plugin-Ardour-Hydrogen no issue. But when I load another Carla Plugin it crashes:
Jack: JackClient::SetupDriverSync driver sem in flush mode
Jack: JackLinuxFutex::Connect name = jack_sem.1000_default_Carla
Jack: Clock source : system clock via clock_gettime
Jack: JackLibClient::Open name = Carla refnum = 3
CarlaEngineOsc::osc_error_handler_TCP(9904, "cannot find free port", "(null)")
CarlaEngineOsc::osc_error_handler_UDP(9904, "cannot find free port", "(null)")
Jack: jack_set_thread_init_callback ext_client 3722770 client 3722770 
Jack: jack_get_uuid_for_client_name ext_client 3722770 client 3722770 
Jack: jack_set_property ext_client 3722770 client 3722770 
Jack: jack_set_property ext_client 3722770 client 3722770 
Jack: JackClient::Activate
Jack: JackPosixThread::StartImp : create non RT thread
Jack: JackPosixThread::ThreadHandler : start
Jack: JackClient::kBufferSizeCallback buffer_size = 1024
Jack: JackClient::Init calling client thread init callback
Jack: JackClient::Init : period = 23219 computation = 100 constraint = 23219
Jack: JackPosixThread::AcquireRealTimeImp priority = 5
Jack: JackClient::ClientNotify ref = 3 name = 8589934595 notify = 19
Jack: JackClient::ClientNotify ref = 3 name = 8589934595 notify = 19
Jack: JackClient::ClientNotify ref = 3 name = Carla notify = 2
Jack: JackClient::kActivateClient name = Carla ref = 3 
Jack: JackClient::Init calling client thread init callback
Jack: JackClient::ClientNotify ref = 3 name = Carla notify = 18
Jack: JackClient::ClientNotify ref = 3 name = Carla notify = 18
Speicherzugriffsfehler (Speicherabzug geschrieben)

**Comment by wundermusiker** [comment]
Ardour gets unsuable as well. So might affect Jack/Cadence as well.
Using Carla version 2.0.96 (2.1-RC1)
  Python version: 3.6.9
  Qt version:     5.10.1
  PyQt version:   5.10.1
  Binary dir:     /usr/lib/carla
  Resources dir:  /usr/share/carla/resources
jack2 version 1.9.14
Cadence 0.9.1

**Comment by wundermusiker** [comment]
With all the other applications seem to be fine, I guess it is linked to Hydrogen 1.0.0 causing the crash.

**Comment by RobinHerbots** [comment]
@falkTX , 

Ok it seems I didn't install the carla bridges.  More plugins are working now without crashing carla.

Still dexed and tunefish are just crashing carla.

**Comment by falkTX** [comment]
I need to install 19.10 in a VM to try to reproduce.

**Comment by falkTX** [comment]
I did just that, got no crash with either carla or carla-git.
19.10 with kx repos added, installed carla and vst plugins from the kx repos.

So I have a few questions...

1. are you using jack?
2. what "engine mode" is carla running on?
3. the plugins you are loading, where are they from?

**Comment by RobinHerbots** [comment]
> * are you using jack?  **Yes**
> * what "engine mode" is carla running on? **I'm just trying out, so I gues it is patchbay**
> * the plugins you are loading, where are they from?
**I downloaded dexed from https://github.com/asb2m10/dexed/releases**  (linux version)

I must say that the windows vsts I tried all seem to work after installing the bridges.  I can connect my controller and play.  I currently try carla on standalone before connecting to a DAW.
Maybe I can try to install the vst plugins from the kx repos also.

**Comment by falkTX** [comment]
I used that build in carla inside 19.10 VM and it loads fine here. :/

![Screenshot_20200129_131415](https://user-images.githubusercontent.com/1334853/73359651-413a7100-4299-11ea-91e7-8dc7cee128ed.png)

**Comment by RobinHerbots** [comment]
@falkTX, 
Yes that looks nice ;-)
You now a way how I can figure why this is happening on my system, besides the few log messages from carla.
I will first try the vsts from the kx repos and see if they work.  I will let you know this evening.

I also want to try out the non -daw which looks promising, but that is another story.

**Comment by falkTX** [comment]
Testing to see if vsts from the kxstudio repos crash or not will be a first good start.
The problem might be in the vsts from dexed and other sources. The plugin authors do not always build for linux in the proper way.

One thing that helps debugging here a bit is to disable the logs tab in carla (first page in the settings), so that we see the logs in the terminal instead.
This way if there is a final message just before the plugin crashes the entire thing, we will see it.

Do you have any of the experimental settings enabled?
Please try disabling it all and try to load dexed again.

Another thing to try is running carla under a debugger.
Install `gdb` and run `carla --gdb` to start that.
then the usual debugging stuff applies, here we just care to get a bit of a backtrace.
basically running `r` in the debugger, then triggering the crash, and then running `bt` (r and bt are shortnames for gdb commands)
the output of `bt` is what matters here.
carla kxstudio build has no debug symbols, but at least we should be able to see if the crash is something in carla or the plugin.

**Comment by RobinHerbots** [comment]
@falkTX , 

Ok thx for the info.
I will test it this evening, as soon I have access to my pc.

**Comment by RobinHerbots** [comment]
@falkTX , 

I first tried installing carla with dexed from github on another pc with ubuntu 19.10 and it all works fine. 

I installed some plugins including dexed from the kx repos on my main pc and they also all work including dexed.

But still on the same machine when loading dexed from the github repo I get the following.

**From console**
```
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
Carla 2.0.96 (2.1-RC1) started, status:
  Python version: 3.7.5
  Qt version:     5.12.4
  PyQt version:   5.12.3
  Binary dir:     /usr/lib/carla
  Resources dir:  /usr/share/carla/resources
qt.svg: <input>:552:178: Could not resolve property: #linearGradient5167
qt.svg: <input>:30274: Could not resolve property: #linearGradient5167
qt.svg: <input>:30274: Could not resolve property: #pattern5614
qt.svg: <input>:30274: Could not resolve property: #pattern5626
qt.svg: <input>:5672: Could not resolve property: #radialGradient3709
qt.svg: <input>:3226: Could not resolve property: #linearGradient3893-1
qt.svg: <input>:3190: Could not resolve property: #linearGradient3893-1
qt.svg: <input>:3208: Could not resolve property: #linearGradient3893-1
libjack.so.0 loaded successfully!
qt.svg: <input>:684: Could not resolve property: #pattern5614
qt.svg: <input>:684: Could not resolve property: #pattern5626
qt.svg: link #XMLID_34_ is undefined!
qt.svg: link #XMLID_34_ is undefined!
qt.svg: link #XMLID_36_ is undefined!
Illegal instruction (core dumped)
```

**From gdb**
```
Thread 1 "carla" received signal SIGILL, Illegal instruction.
0x00007fff88b774c5 in Tanh::init() () from /home/robin/VST/Dexed/Dexed.so
(gdb) bt
#0  0x00007fff88b774c5 in Tanh::init() () from /home/robin/VST/Dexed/Dexed.so
#1  0x00007fff88b56e10 in DexedAudioProcessor::DexedAudioProcessor() ()
   from /home/robin/VST/Dexed/Dexed.so
#2  0x00007fff88b570f0 in createPluginFilter() ()
   from /home/robin/VST/Dexed/Dexed.so
#3  0x00007fff889c4505 in createPluginFilterOfType(juce::AudioProcessor::WrapperType) () from /home/robin/VST/Dexed/Dexed.so
#4  0x00007fff889a85ca in VSTPluginMain () from /home/robin/VST/Dexed/Dexed.so
#5  0x00007fffd6b268be in ?? () from /usr/lib/carla/libcarla_standalone2.so
#6  0x00007fffd6b2f07e in CarlaBackend::CarlaPlugin::newVST2(CarlaBackend::CarlaPlugin::Initializer const&) () from /usr/lib/carla/libcarla_standalone2.so
#7  0x00007fffd6a845e3 in CarlaBackend::CarlaEngine::addPlugin(CarlaBackend::BinaryType, CarlaBackend::PluginType, char const*, char const*, char const*, long, void const*, unsigned int) () from /usr/lib/carla/libcarla_standalone2.so
```

**Comment by falkTX** [comment]
Alright, that is a lot more clear.
`SIGILL` often happens when the binary tries to do an operation not supported in the current CPU. for example, a plugin that is compiled with SSE4 but the CPU does not support such instructions.

There is nothing wrong in Carla then, the issue is with the plugin developers not building a proper binary (Juce has a terrible habit of using `-march=native` by default on Linux makefiles)
There is also nothing that Carla can do about this, the plugin is basically "broken" here.

All is fine using the kxstudio repository binaries, so I recommend to use those.
If there is an opensource plugin that you really want to use but it is not yet in there, let me know.
I am closing this ticket now since it is not Carla's fault.

**Comment by RobinHerbots** [comment]
@falkTX , 

Yes it is obviously not carla's fault.  There are already many plugins available in de kx repos, so I will first explore them.

Maybe as a last remark, can't you catch the exception, so carla doesn't crash caused by the plugin.   Showing a message that the plugin itself crashed would show Carla as stable and can save you and the enduser a lot of time investigating.

I want to thank you for your time and support!  That is really appreciated!

Kind regards, 
Robin

**Comment by falkTX** [comment]
> Maybe as a last remark, can't you catch the exception, so carla doesn't crash caused by the plugin. Showing a message that the plugin itself crashed would show Carla as stable and can save you and the enduser a lot of time investigating.

This would be quite nice, but it is also very tricky.
Renoise is able to do it though, so it should be possible... But first I need a way to replicate the issue.

**Comment by RobinHerbots** [comment]
@falkTX , 

Can this help? https://www.codeproject.com/Articles/207464/Exception-Handling-in-Visual-Cplusplus

If I find some time this weekend I will try to compile Carla and see if I can catch the exception.  The error appears on my pc so I have a good testcase.  

I'm not really at home with developing on linux.  Do you use some IDE?

**Comment by falkTX** [comment]
Something like that, but in a way that works for Linux/Unix. So instead of being catching exceptions globally, we do it via signal handling.
It is quite tricky, since it is a global operation. So for one it cannot be done when running as plugin, only in standalone mode.
I have some ideas on how to do it. but main issue now is finding a way to reproduce.

**Comment by RobinHerbots** [comment]
You can get the dexed source code.  Adjust to trigger a sigill.  And use that as test.

The add plugin code is in https://github.com/falkTX/Carla/blob/develop/source/backend/engine/CarlaEngine.cpp.
Isn't it sufficient to handle the signal in there?  
Where can I get some info about signal handling in linux?

**Comment by RobinHerbots** [comment]
Ok after reading a bit, ...  so you register the signal handler globally, ....  somewhere at the start of your application.  When the signals happens you can act accordingly.  Why can this only be done in standalone and not when running as plugin?  Because when you handle the signal the main application isn't aware?  Isn't there a mechanism as rethrow (resignal)?

**Comment by falkTX** [comment]
It is just not nice to mess up with global things when running as a plugin.
You never know how the host expect things to work, modifying that state will affect not just carla as plugin, but all other plugins.

**Comment by RobinHerbots** [comment]
Yes I agree, but maybe a gradually mess can be usefull.  Only handle a specific case.  
I see it as the signal belongs to the scope of Carla it should be handled by Carla, otherwise you should leave it up to the host.  Offcourse, I'm just thinking up loud, but I don't know this is possible or easy to do.
